dask_deltatable/write.py

```diff
@@ -0,0 +1,262 @@
+    file_options: ds.ParquetFileWriteOptions | None = None,
+    max_partitions: int | None = None,
+    max_open_files: int = 1024,
+    max_rows_per_file: int = 10 * 1024 * 1024,
+    min_rows_per_group: int = 64 * 1024,
+    max_rows_per_group: int = 128 * 1024,
+    name: str | None = None,
+    description: str | None = None,
+    configuration: Mapping[str, str | None] | None = None,
+    overwrite_schema: bool = False,
+    storage_options: dict[str, str] | None = None,
+    partition_filters: list[tuple[str, str, Any]] | None = None,
+):
+    """Write a given dask.DataFrame to a delta table
+
+    TODO:
+    """
+    table, table_uri = try_get_table_and_table_uri(table_or_uri, storage_options)
+
+    # We need to write against the latest table version
+    if table:
+        table.update_incremental()
+
+    __enforce_append_only(table=table, configuration=configuration, mode=mode)
+
+    if filesystem is None:
+        if table is not None:
+            storage_options = table._storage_options or {}
+            storage_options.update(storage_options or {})
+
+        filesystem = pa_fs.PyFileSystem(DeltaStorageHandler(table_uri, storage_options))
+
+    if isinstance(partition_by, str):
+        partition_by = [partition_by]
+
+    if table:  # already exists
+        if schema != table.schema().to_pyarrow() and not (
+            mode == "overwrite" and overwrite_schema
+        ):
+            raise ValueError(
+                "Schema of data does not match table schema\n"
+                f"Table schema:\n{schema}\nData Schema:\n{table.schema().to_pyarrow()}"
+            )
+
+        if mode == "error":
+            raise AssertionError("DeltaTable already exists.")
+        elif mode == "ignore":
+            return
+
+        current_version = table.version()
+
+        if partition_by:
+            assert partition_by == table.metadata().partition_columns
+        else:
+            partition_by = table.metadata().partition_columns
+
+        if table.protocol().min_writer_version > MAX_SUPPORTED_WRITER_VERSION:
+            raise DeltaProtocolError(
+                "This table's min_writer_version is "
+                f"{table.protocol().min_writer_version}, "
+                "but this method only supports version 2."
+            )
+    else:  # creating a new table
+        current_version = -1
+
+    # FIXME: schema is only known at this point if provided by the user
+    if partition_by and schema:
+        partition_schema = pa.schema([schema.field(name) for name in partition_by])
+        partitioning = ds.partitioning(partition_schema, flavor="hive")
+    else:
+        if partition_by:
+            raise NotImplementedError("Have to provide schema when using partition_by")
+        partitioning = None
+    if mode == "overwrite":
+        # FIXME: There are a couple of checks that are not migrated yet
+        raise NotImplementedError()
+
+    written = df.map_partitions(
+        _write_partition,
```

Comment:
Why can't we use `write_deltalake` from `deltatable` for writing a partition?

Reply:
We want to separate the writing of the parquet files from the commit to the log. The entire `to_deltalake` is supposed to be one transaction. Using the `deltatable.write_deltalake` would create a commit per partition. This is not only not what a transaction is supposed to be but this would almost guarantee that we'd have write conflicts due to concurrent writes.